### ISSUE TITLE
Low mAP on Pascal VOC2007 dataset.

### ISSUE THREAD

**Issue by Scofieldtangsiwei** [kevinjliang/tf-Faster-RCNN#27]
I followed the instruction and trained a model based on pre-trained resnet50 v1 model with Pascal VOC2007 dataset. But the mAP was always stuck at 30 or so. Has anyone tried this dataset yet? Any help will be greatly appreciated!

**Comment by paviddavid** [comment]
@Scotfieldtangsiwei: I got totally the same behavior! 

Did you find a solution for the low mAP? 
@kevinjliang Any ideas?

**Comment by Wu-Xiuchao** [comment]
@Scofieldtangsiwei  I got 54% mAP. Still lower than what  author's paper showed. I used VGG16 and the dataset was voc2007 train-val.
Actually I  want to know if it is because the way we implement roi-pooling that caused this problem.

**Comment by paviddavid** [comment]
@Wu-Xiuchao I use the ResNet v1 50 and the train-val voc2007 data , trained for 40 epochs and around epoch 20 the mAP reaches 31% - 32% and stagnates. I do not change the parameters in pascal_voc2007.yml and I do not know why the mAP is low.

**Comment by Wu-Xiuchao** [comment]
@paviddavid Hi, I think there is no need to train 40 epoches, 15 or less is enough. I just use the code of roi-pooling of @kevinjliang ,and the rest of the code refer to [smallcorgi/Faster-RCNN_TF](https://github.com/smallcorgi/Faster-RCNN_TF)

**Comment by Scofieldtangsiwei** [comment]
You guys should check the size of the feature map output by your backbone net. I found the author forget to get rid of the last average pooling in the resnet.

在 2019年4月3日，04:06，Xiuchao Wu <notifications@github.com<mailto:notifications@github.com>> 写道：


@paviddavid<https://github.com/paviddavid> Hi, I think there is no need to train 40 epoches, 15 or less is enough. I just use the code of roi-pooling of @kevinjliang<https://github.com/kevinjliang> ,and the rest of the code refer to smallcorgi/Faster-RCNN_TF<https://github.com/smallcorgi/Faster-RCNN_TF>

**Comment by paviddavid** [comment]
@Scofieldtangsiwei Thanks for your remark. However, there is a point I do not understand. How should I get rid of the average pooling layer in the resnet? How have you solved this issue?
I would appreciate every help because I am new to this subfield.

**Comment by Scofieldtangsiwei** [comment]
The final feature map size should be 1/16 of the original image. You can check the resnet implementation, if the total stride is 32, you should change it to 16.

在 2019年4月3日，04:17，David <notifications@github.com<mailto:notifications@github.com>> 写道：


@Scofieldtangsiwei<https://github.com/Scofieldtangsiwei> Thanks for your remark. However, there is a point I do not understand. How should I get rid of the average pooling layer in the resnet? How have you solved this issue?
I would appreciate every help because I am new to this subfield.

**Comment by paviddavid** [comment]
@Scofieldtangsiwei So, you are talking about the `_feat_stride` in the `faster_rcnn_resnet50ish.py` file? As you can see in the attached screenshot, the stride is 16, not 32.  Or do you mean I should change something within `resnet50_reduced.py` ?

![image](https://user-images.githubusercontent.com/45428272/55468613-1002fd00-5604-11e9-82bb-26b020ece07f.png)

Thanks in advance

**Comment by Scofieldtangsiwei** [comment]
I don’t think the dimension in the screen shot can indicate the feat stride of your backbone net, since the height and width are all “?” in tensorflow. But you can check the resnet.py, the default _feat_stride is 2^5. And you can also check the implementation of the class resnet, the total stride is 32 rather than 16.

在 2019年4月3日，04:32，David <notifications@github.com<mailto:notifications@github.com>> 写道：


@Scofieldtangsiwei<https://github.com/Scofieldtangsiwei> So, you are talking about the _feat_stride in the faster_rcnn_resnet50ish.py file? As you can see in the attached screenshot, the stride is 16, not 32. Or do you mean I should change something within the downloaded weights of the ResNet?

[image]<https://user-images.githubusercontent.com/45428272/55468613-1002fd00-5604-11e9-82bb-26b020ece07f.png>

Thanks in advance

**Comment by paviddavid** [comment]
Ok, in `resnet.py` I can find `_feat_stride = 2**5` as you mentioned. However, in `faster_rcnn_resnet50ish.py` the `resnet50.py` is not loaded but the `resnet50_reduced.py` and therefore, only the resnet50_reduced network is used, or? 
In `faster_rcnn_resnet50ish.py` in line 95 the `feature_maps = resnet50_reduced(x)` and line 97 `_feat_stride = 16` indicate that the correct total stride is per default loaded, am I right or do I miss something? 
Hence, the implementation should already use the right value?

**Comment by Scofieldtangsiwei** [comment]
I think it is where the author gets wrong. In the resnet50_reduced.py, you can find a resnet50_reduced function. In that function, the code defines the structure of resnet 50. However, the block3 should have stride 1 rather than stride 2. In your code, during training, you can check the size of your original image and the size of the final feature map output by the resnet using sess.run(). I believe you will get what I mean.

**Comment by paviddavid** [comment]
Hi @Scofieldtangsiwei, thanks for the advice. The accuracy went up using stride=1 instead of stride=2. Now, I get 43% (before: 33%) after 140 epochs (around 22 hours of training). This is already good, but in similar papers, they achieve an accuracy around 65-70% on the Pascal_VOC2007 dataset. Do you know why there is this large gap in the accuracy?

**Comment by Scofieldtangsiwei** [comment]
Sorry, I couldn't debug the model to reach the Pascal VOC baseline and finally changed for another Faster RCNN model.
